troposphere/static/js/components/admin/ResourceRequest/AllocationSourceView.jsx

```diff
@@ -0,0 +1,186 @@
+
+    onInputChange({ target: { value }}) {
+        let { allocationSource } = this.state;
+        allocationSource.set("compute_allowed", +value);
+        this.setState({ allocationSource });
+    }
+
+    renderEditableField(num) {
+        let style = {
+            margin:"0 5px"
+        }
+        return (
+        <input style={style}
+               onChange={this.onInputChange.bind(this)}
+               type="number"
+               value={num} />
+        );
+    }
+
+    onSave() {
+       let { allocationSource, cancellables } = this.state;
+
+        // Create cancellable versions of callbacks to prevent calling
+        // setState on unmount
+        let onSuccess = cancellable(
+            allocation => {
+                allocationSource.set(allocation);
+                this.setState({ allocationSource, isPending: false, isEditing: false });
+            }
+        );
+        let onErr = cancellable(() => this.setState({ isPending: false }));
+        cancellables.push(onSuccess);
+        cancellables.push(onErr);
+
+        this.setState({ isPending: true });
+        // Clone our allocation, so that call to props.onSave doesn't affect
+        // our state
+        this.props.onSave(allocationSource.clone())
+            .then(onSuccess, onErr);
+    }
+
+    renderSaveText() {
+        let { isPending } = this.state;
+        let buttonText = isPending ? "loading" : "save";
+        return <span> <a style={{ float: "right" }} onClick={ this.onSave.bind(this) }>{ buttonText }</a></span>;
+    }
+
+    render() {
+        let { allocationSource: cloudAllocation } = this.props;
+        let { isEditing, isPending, allocationSource } = this.state;
+        let { percent, total } = this.getUsage(allocationSource);
+        let { name } = this.props.allocationSource.pick("name");
+        let editableTotal =
+            isEditing
+            ? this.renderEditableField(total)
+            : total;
+
+        let hasChanged =
+            // Check if the cloud allocation and local allocation differ
+            allocationSource.keys().some(attr => allocationSource.get(attr) !== cloudAllocation.get(attr));
+
+        let statusElement = null;
+        if (isPending) {
+            statusElement = <span style={{ float: "right", margin: 0 }} className="loading-small"></span>;
+        } else if (hasChanged) {
+            statusElement = <span> <a style={{ float: "right" }} onClick={ this.onSave.bind(this) }>save</a></span>;
+        }
+
+        let allocationLabel = (
+        <span>
+             { `${percent}% used of ` }
+             { editableTotal }
+             { ` Allocation from ${name}` }
+             { this.renderEditIcon() }
+             { statusElement }
+        </span>
+        );
+
+        return (
+        <ProgressBar startColor="#56AA21" startValue={percent} label={allocationLabel} />
+        );
+    }
+}
+
+AllocationEditableSaveableView.propTypes = {
+    onSave: React.PropTypes.func.isRequired
+}
```

Comment:
First time i've seen this syntax used to define `propTypes` for a class, could you comment on this new usage?

Reply:
propTypes no longer exist as part of React, but we're using an addon to keep the functionality. When you use es6 classes, propTypes are sought as an attribute on the class.

Reply:
@steve-gregory - 

The quick answer is: 

Since `AllocationEditableSaveableView` is defined as an ES6 (aka ES2015) class (instead of `React.createClass({...})`), this is how `propTypes` are defined (given you're not passing in the object literal to `createClass(..)`.

Also, `propTypes` has moved into it's own module (out of the `React` module). So there is a full blown "module" called `PropTypes`. Facebook has provided a codemod that will do some of the translation. 

Why extend `React.Component`? 

The traditional inheritance that Connor leverages here is quite elegant (in my opinion). So, I think this is a great example of where you'd use this over a regular prototypal definition.